addons/website/models/website_visitor.py

```diff
@@ -307,11 +307,27 @@ def _link_to_visitor(self, target):
 
         return target
 
-    def _cron_archive_visitors(self):
+    def _cron_clean_inactive_visitors(self):
+        """ Unlink inactive visitors (see '_inactive_visitors_domain' for details).
+
+        Visitors were previously archived but we came to the conclusion that archived visitors
+        have very little value and bloat the database for no reason. """
+
+        inactive_visitors = self.env['website.visitor'].sudo() \
```

Comment:
Hello @nseinlet 

Question about this task which aims to clean useless website.visitors records.
The current code is using a simple approach: "archive inactive website.visitors" (not logged-in for more than 30 days).

The new approach will DELETE inactive visitors, because it does not make much sense to keep those forever as archived, letting the table grow rapidly and infinitely.

But it will make more checks to ensure that the records don't have side-records attached to them.

So it will look like the following for example with crm and website_event_track installed:
```sql
SELECT website_visitor.id
FROM website_visitor
WHERE 
(website_visitor.last_connection_datetime < '2022-02-08 09:59:14.904120'
 AND website_visitor.partner_id IS NULL)
AND
(website_visitor.id not in
    (SELECT website_visitor_id
	 FROM crm_lead_website_visitor_rel
	 where website_visitor_id is not null)
)
AND
(website_visitor.id not in
	(SELECT visitor_id
	 FROM event_registration
	 where visitor_id is not null)
)
AND 
(website_visitor.id not in
	(SELECT visitor_id
	 FROM event_track_visitor where visitor_id is not null)
)
ORDER BY  website_visitor.last_connection_datetime DESC
```
(And then `unlink()`)

Which will check that the website.visitor is not attached to leads, event registrations and event favorited tracks before deleting.
(More apps add their own conditions, but I think you get the idea).

Should we add a migration script step that cleans the table before running this CRON, based on conditions above?
(It will be a bit tricky to do because based in installed modules, but it could be done).

Or can we let the CRON clean the whole table at its first run?
I know this table is very large so I prefer asking your opinion to avoid any issues.

Thanks for your help.

cc @tde-banana-odoo

Reply:
Let the cron do the job.
```sql
AND 
(website_visitor.id not in
	(SELECT visitor_id
	 FROM event_track_visitor where visitor_id is not null)
)
```
Is better in this form
```sql
AND NOT EXISTS (SELECT 1 FROM event_track_visitor WHERE visitor_id=website_visitor.id)
```

Reply:
@nseinlet The request I referenced here is actually the one built by the ORM using leaves with `('xxx_ids', '=', False)`.

Building a regular domain with `_inactive_visitors_domain` makes the code much cleaner, maintainable, and most importantly modular since other modules are adding their criteria into the query.

We could attempt building raw SQL but it would not be clean and probably hard to use.

Does it matter a lot in term of query optimization?
Will this still perform sufficiently?

Again, thanks for the help!